### PR TITLE
syncer: fix generated column in where condition

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve? 

Currently we remove all generated column for DML, but generated column in where condition is valid.

### What is changed and how it works?
- keep generated column in index searching
- keep generated column in where condition when generating `update`/`delete` DML

### Check List 

Tests 

 - Integration test

Side effects

 - Increased code complexity

### CODE REVIEW

**Comment by amyangfei** [comment]
PTAL @GregoryIan @csuzhangxc

File: syncer/dml.go [hunk 1]
```diff
@@ -139,17 +140,21 @@ func genUpdateSQLs(schema string, table string, data [][]interface{}, columns []
...
+		for i := range oriOldData {
+			oriOldValues = append(oriOldValues, castUnsigned(oriOldData[i], originalColumns[i].unsigned, originalColumns[i].tp))
+		}
 
 		if len(defaultIndexColumns) == 0 {
-			defaultIndexColumns = getAvailableIndexColumn(indexColumns, oldValues)
+			defaultIndexColumns = getAvailableIndexColumn(originalIndexColumns, oriOldValues)
 		}
 
 		ks := genMultipleKeys(columns, oldValues, indexColumns)
```
Comment (GregoryIan) [inline syncer/dml.go]: should we use `oriOldValues`?

**Comment by GregoryIan** [comment]
Rest LGTM

**Comment by amyangfei** [comment]
PTAL @GregoryIan

File: syncer/dml.go [hunk 2]
```diff
@@ -74,24 +83,31 @@ func (c *GenColCache) clearTable(schema, table string) {
...
+		originalIndexColumns = param.originalIndexColumns
+	)
 	sqls := make([]string, 0, len(dataSeq))
 	keys := make([][]string, 0, len(dataSeq))
 	values := make([][]interface{}, 0, len(dataSeq))
 	columnList := genColumnList(columns)
 	columnPlaceholders := genColumnPlaceholders(len(columns))
-	for _, data := range dataSeq {
+	for dataIdx, data := range dataSeq {
 		if len(data) != len(columns) {
```
Comment (GregoryIan) [inline syncer/dml.go]: also check `originalColumns ` and `originalDataSeq `?

File: syncer/dml.go [hunk 3]
```diff
@@ -140,16 +177,30 @@ func genUpdateSQLs(schema string, table string, data [][]interface{}, columns []
 			changedValues = append(changedValues, castUnsigned(changedData[i], columns[i].unsigned, columns[i].tp))
 		}
 
+		oriOldValues := make([]interface{}, 0, len(oriOldData))
```
Comment (GregoryIan) [inline syncer/dml.go]: `genUpdateSQLs`, `genInsertSQLs`` has these codes in same pattern:

```
...values
for i := range data {
    values = append(values,....)
}

if len(columns) == len(originalColumns) {
  originalVals = values
} else {
    for {....}
}
```

can we extract it a function?

File: syncer/dml.go (hunk 2 above)
Comment (amyangfei) [reply syncer/dml.go]: not needed, data/columns and origianlData/originalColumns have the same length distance ?

Comment (amyangfei) [reply syncer/dml.go]: add column and data length check in generated column prune

**Comment by GregoryIan** [comment]
LGTM

**Pull request merged**